### PR TITLE
Add more functionalities to client4.js

### PR DESCRIPTION
**Pull request by xuxip**
Below Functionalities are in Mattermost API v4 but missing in client4.js:
- patchUser (this partially update an user so that developers don't have specify every field in the user object)
- getTeamByName (this would be very helpful when team id is unknown)
- getChannelByNameAndTeamName (this would be very helpful when only team name and channel name are given)
- deleteTeam

I really hope these features to be included in the driver and benefit the future users.

### CODE REVIEW

**Comment by enahum** [comment]
Hi @xuxip thanks for the PR, can I ask you to please sign the CLA ?

**Comment by xuxip** [comment]
Hi @enahum, I signed the CLA.

**Comment by xuxip** [comment]
@enahum, track event is added. What did you mean to include action and tests?

**Comment by enahum** [comment]
I meant that every client call has a redux action that in turn will populate the reducer and that should have their unit tests as well.

You can always use the client driver without using the action.

**Comment by xuxip** [comment]
How to run the tests?

**Comment by enahum** [comment]
You can run the test by doing a `make test` in the terminal

**Comment by xuxip** [comment]
@enahum, I included the action and test for the two functionality I added.

File: src/actions/users.js [hunk 1]
```diff
@@ -937,6 +937,40 @@ export function updateMe(user) {
...
+        }
+
+        dispatch(batchActions([
+            {type: UserTypes.RECEIVED_PROFILE, data},
+            {type: UserTypes.UPDATE_USER_SUCCESS}
+        ]), getState);
+
+        return {data};
+
+        // const actions = [
```
Comment (enahum) [inline src/actions/users.js]: can you please remove the commented code?

**Comment by xuxip** [comment]
@enahum Do you find any other problems?

File: src/actions/users.js [hunk 2]
```diff
@@ -937,6 +937,27 @@ export function updateMe(user) {
     };
 }
 
+export function patchUser(userId, user) {
```
Comment (enahum) [inline src/actions/users.js]: Why are we passing the userId as an argument? I think the user object should have the id

File: src/client/client4.js [hunk 3]
```diff
@@ -285,6 +285,15 @@ export default class Client4 {
         );
     }
 
+    patchUser = async (userId, user) => {
```
Comment (enahum) [inline src/client/client4.js]: same as above

File: src/actions/users.js (hunk 2 above)
Comment (xuxip) [reply src/actions/users.js]: Yes! You are right. Fixed

**Comment by xuxip** [comment]
@enahum I added getChannelByNameAndTeamname https://github.com/mattermost/mattermost-redux/pull/353/commits/ee52feadf361b009c4bc2445dc1e04b96dca836c

**Comment by jwilander** [comment]
Thanks @xuxip, would you be able to add the action and test for `getChannelByNameAndTeamname` as well? That would be super helpful

**Comment by enahum** [comment]
And please try and add different functionalities in different PR's I'm seeing a constant push of commits to this PR with more and more functionality being added

**Comment by xuxip** [comment]
@enahum @jwilander I made following changed to create deleteTeam in action and action test, but the test keeps failing and got the error ''TypeError: Cannot read property 'status' of undefined"
**test/actions/teams.test.js**
`    it('deleteTeam', async () => {
        const secondClient = TestHelper.createClient4();

        nock(Client4.getUsersRoute()).
            post('').
            query(true).
            reply(201, TestHelper.fakeUserWithId());

        const user = await TestHelper.basicClient4.createUser(
            TestHelper.fakeUser(),
            null,
            null,
            TestHelper.basicTeam.invite_id
        );

        nock(Client4.getUsersRoute()).
            post('/login').
            reply(200, user);
        await secondClient.login(user.email, 'password1');

        nock(Client4.getTeamsRoute()).
            post('').
            reply(201, TestHelper.fakeTeamWithId());
        const secondTeam = await secondClient.createTeam(
            TestHelper.fakeTeam());

        nock(Client4.getTeamsRoute()).
            delete(`/${secondTeam.id}`).
            reply(200, OK_RESPONSE);

        await Actions.deleteTeam(
            secondTeam.id
        )(store.dispatch, store.getState);

        const deleteRequest = store.getState().requests.teams.deleteTeam;

        if (deleteRequest.status === RequestStatus.FAILURE) {
            throw new Error(JSON.stringify(deleteRequest.error));
        }

        const {teams, myMembers} = store.getState().entities.teams;
        assert.ifError(teams[secondTeam.id]);
        assert.ifError(myMembers[secondTeam.id]);
    });`

**actions/teams.js**
`export function deleteTeam(teamId) {
    return async (dispatch, getState) => {
        dispatch({type: TeamTypes.DELETE_TEAM_REQUEST}, getState);

        let deleted = null;
        try {
            deleted = await Client4.deleteTeam(teamId);
        } catch (error) {
            forceLogoutIfNecessary(error, dispatch);
            dispatch(batchActions([
                {type: TeamTypes.DELETE_TEAM_FAILURE, error},
                logError(error)(dispatch)
            ]), getState);
            return {error};
        }

        dispatch(batchActions([
            {
                type: TeamTypes.RECEIVED_TEAM_DELETED,
                data: {id: teamId}
            },
            {
                type: TeamTypes.DELETE_TEAM_SUCCESS
            }
        ]), getState);

        return {data: deleted};
    };
}`

**action_types/teams.js**
``

**Comment by xuxip** [comment]
@enahum @jwilander I added actions and tests for all the functionalities that I have added. Please let me know if you have any questions. Wish you have a Happy New Year!

**Comment by xuxip** [comment]
@enahum @jwilander Did you find any other problems in this PR?

**Comment by coreyhulen** [comment]
Hi @xuxip,

Sorry about the long delay.  Totally our fault here, looks like it might have fallen through the cracks after the new year.  We'll get it queued up and reviewed ASAP.

**Pull request merged**